src/libos/src/vm/vm_area.rs

```diff
@@ -68,10 +154,181 @@ impl VMArea {
         &self.range
     }
 
+    fn pages(&self) -> &PageTracker {
+        debug_assert!(!self.is_fully_committed());
+        self.pages.as_ref().unwrap()
+    }
+
+    fn pages_mut(&mut self) -> &mut PageTracker {
+        debug_assert!(!self.is_fully_committed());
+        self.pages.as_mut().unwrap()
+    }
+
     pub fn pid(&self) -> pid_t {
         self.pid
     }
 
+    pub fn initializer(&self) -> &Option<VMInitializer> {
+        &self.initializer
+    }
+
+    pub fn is_reserved_only(&self) -> bool {
+        if let Some(pages) = &self.pages {
+            return pages.is_reserved_only();
+        } else {
+            false
+        }
+    }
+
+    pub fn is_fully_committed(&self) -> bool {
+        self.pages.is_none()
+    }
+
+    pub fn is_partially_committed(&self) -> bool {
+        if let Some(pages) = &self.pages {
+            return pages.is_partially_committed();
+        } else {
+            false
+        }
+    }
+
+    pub fn init_memory(mut self, options: &VMMapOptions) -> Result<Self> {
+        trace!("init_memory");
+        let mut vm_area = self;
+        let page_policy = options.page_policy();
+
+        // Commit pages if needed
+        if !vm_area.is_fully_committed() && page_policy == &PagePolicy::CommitNow {
```

Comment:
if vm_area.is_partially_committed() {
            if page_policy == &PagePolicy::CommitNow {
                vm_area.pages_mut().commit_current_vma(true)?;
                vm_area.pages = None;
            } else {
                return vm_area.init_committed_memory().map(|_| vm_area)
            }
        }
        // Initialize the memory of the fully new range
        unsafe {
            trace!("vma is fully committed");
            let buf = vm_area.range().as_slice_mut();
            options.initializer().init_slice(buf)?;
        }

        // Set memory permissions
        if !options.perms().is_default() {
            vm_area.modify_protection_lazy(
                None,
                VMPerms::DEFAULT,
                vm_area.perms(),
                page_policy == &PagePolicy::CommitNow,
            );
        }

Reply:
Thank you for your review. Here it also needs to consider the vm_area `is_reserved_only`. I changed a little bit of my previous code to return as early as possible.